### DISCUSSION TITLE
What is the correct way to upgrade?

### DISCUSSION THREAD

**Discussion by chall3ng3r** [cockpit-project/cockpit#17800] [Q&A] [answered]
I'm running v264-1~bpo20.04.1, Just hit by the following error:

`This web browser is too old to run the Web Console (missing selector(:is():where()))`

I noticed there's v277 and when I check for updates, I only get the message that my current version is the latest.

The install guide also does not mention the process to update, either via apt or manually. It'd be great if someone could share the simple step by step instructions for performing the update.

Thanks.

**Comment by jelly** [comment]
See https://cockpit-project.org/running.html#ubuntu for our installation instructions for Ubuntu 20.04, it should include installing from backports.

**Comment by chall3ng3r** [comment]
Thanks for the reply. I installed it following the same guide few months ago. Added some additional apps in it as well. 

My question is regarding updating Cockpit to latest version. There is no guide for it, either through apt or manually.

**Reply by garrett** [reply]
The [Debian section](https://cockpit-project.org/running.html#debian) states:

"When updating Cockpit-related packages and any dependencies, make sure to use -t ...-backports as above, so backports are included."

The Ubuntu section should also say something similar... but doesn't.

I've opened an issue @ https://github.com/cockpit-project/cockpit-project.github.io/issues/566

**Reply by garrett** [reply]
Fixed with PR https://github.com/cockpit-project/cockpit-project.github.io/pull/567

The backports upgrade note is now on the website too @ https://cockpit-project.org/running.html#ubuntu

Thanks for pointing it out!

**Comment by chall3ng3r** [comment]
Thanks for the reply. 

I've tried with the -t option, it still says I'm on the latest version. How do I actually get the latest v277? I'm on v264.

**Reply by garrett** [reply]
Oh. I see the same thing here in Ubuntu 20.04 LTS. @martinpitt should know more.

Cockpit from Ubuntu 20.04 LTS, without using backports (from `dpkg -l cockpit`):

```
ii  cockpit        215-1        all          Web Console for Linux servers
```

Cockpit from the backports, I get this:

```
ii  cockpit        264-1~bpo20.04.1 all          Web Console for Linux servers
```

It looks like Ubuntu 20.04 LTS doesn't have a build with a fix?

**Reply by chall3ng3r** [reply]
I'm running the same version of Ubuntu, and I get exactly same Cockpit versions. I hope there's some kind of script I can run to force the update directly from source rather than wait for the apt repo.

**Reply by garrett** [reply]
There's a tracking bug for the official package @ https://bugs.launchpad.net/ubuntu/+source/cockpit/+bug/1990623 

- It still has "Confirmed" status for "Focal" (20.04).
- "Jammy" (22.04 LTS) is still stuck at "In Progress".
- "Kinetic" (22.10) has a release.

@martinpitt has a testing repo for 22.04, but not 20.04 @ https://launchpad.net/~pitti/+archive/ubuntu/fixes/

Apologies; this still seems to not have have been resolved, somehow.

**Reply by garrett** [reply]
You shouldn't have to run a script; the packages should be updated.

The fix is in `login.js` @ https://github.com/cockpit-project/cockpit/commit/ce351ca7079ba447c44796ed179734833aaa6a59

It needs those two asterisks, as browser permitted the previous check but updated `:is()`/`:where()` support where it now fails in newer versions (hence the asterisks, to make them pass again).

---

### Danger area below... not really recommended nor tested.

In Ubuntu 20.04, that file is in the cockpit-ws package.

```
# dpkg -L cockpit-ws | grep login.js
/usr/share/cockpit/static/login.js
```

However, it's minified.

You _could_ theoretically do something like (as root):
```sh
sed -i 's/is():where()/is(*):where(*)/' /usr/share/cockpit/static/login.js
```

But it will be modifying a file that's controlled by a package. And that's generally not a good idea. It might cause problems when upgrading. 

---

Anyway, I hope this is finally fixed everywhere soon. Sorry for the mess.

**Reply by chall3ng3r** [reply]
Thank you for the update. Appreciate it.

I'm currently able to login using Firefox and Chromium as I update them less frequently. Edge (Primary) browser is not able to login.

I hope the bug fix changes are pushed sooner and also the latest v277 version too.

**Reply by martinpitt** [reply]
The released version 215 in Ubuntu 20.04 is actually not affected by this bug. There is currently no fix available for focal-backports, as that is blocked on accepting the fix into jammy-updates in https://launchpad.net/bugs/1990623. It's annoying that apparently SRU processing has been stuck for 6 weeks (!). Once that lands, I'll also backport it to focal (20.04), I can do that myself.

**Reply by garrett** [reply]
I tweeted about it for visibility (just to Ubuntu and Canonical; this isn't for shaming, it's just to highlight it — we're all busy and sometimes things do "slip through the cracks"). I'm certain there are some mutuals between myself and Ubuntu/Canonical who might see it and be able to get the process unstuck.

Hopefully something gets it moving.

**Reply by chall3ng3r** [reply]
> You shouldn't have to run a script; the packages should be updated.
> 
> The fix is in `login.js` @ [ce351ca](https://github.com/cockpit-project/cockpit/commit/ce351ca7079ba447c44796ed179734833aaa6a59)
> 
> It needs those two asterisks, as browser permitted the previous check but updated `:is()`/`:where()` support where it now fails in newer versions (hence the asterisks, to make them pass again).
> 
> ### Danger area below... not really recommended nor tested.
> In Ubuntu 20.04, that file is in the cockpit-ws package.
> 
> ```
> # dpkg -L cockpit-ws | grep login.js
> /usr/share/cockpit/static/login.js
> ```
> 
> However, it's minified.
> 
> You _could_ theoretically do something like (as root):
> 
> ```shell
> sed -i 's/is():where()/is(*):where(*)/' /usr/share/cockpit/static/login.js
> ```
> 
> But it will be modifying a file that's controlled by a package. And that's generally not a good idea. It might cause problems when upgrading.
> 
> Anyway, I hope this is finally fixed everywhere soon. Sorry for the mess.

This worked. 

However, the latest version is still stuck somewhere. I always get v264 when I try to update.

**Reply by garrett** [reply]
> This worked.

Awesome!

> However, the latest version is still stuck somewhere. I always get v264 when I try to update.

Unbelievably and unfortunately, it looks like Ubuntu still hasn't handled their SRU queue since August — https://launchpad.net/ubuntu/jammy/+queue?queue_state=1 — and this includes the Cockpit update with the fix.

**Reply by garrett** [reply]
It's moving along now!

https://bugs.launchpad.net/ubuntu/+source/cockpit/+bug/1990623#yui_3_10_3_1_1666020678203_2318

> Accepted cockpit into jammy-proposed. The package will build now and be available at https://launchpad.net/ubuntu/+source/cockpit/264-1ubuntu0.22.04.1 in a few hours, and then in the -proposed repository.
>
> Please help us by testing this new package. See https://wiki.ubuntu.com/Testing/EnableProposed for documentation on how to enable and use -proposed. Your feedback will aid us getting this update out to other Ubuntu users.
> ...

**Reply by chall3ng3r** [reply]
Thank you. Appreciate it.

I think there should be a self-installable package for Cockpit as well, which users can download and run to manually install without going through update repos.

**Reply by garrett** [reply]
As the packages are not yet officially in Ubuntu's main repo, for sideloading, you can download the proposed packages (which will be what's officially distributed by Ubuntu soon) and install those manually. Or you can download and install the packages from Martin's testing repo. (He's the maintainer of the Ubuntu and Debian packages and has been one of the main official developers of the Cockpit team for years, so the packages are definnitely trustworthy.)

It's also possible to enable these special repos too, but you asked about downloading and installing packages — and it is possible for sure.

As long as the major version is the same, the only package you should need is `cockpit-ws` (as long as the package dependencies are happy with it and don't require a bump too).

**Comment by garrett** [comment]
I was hoping this would be solved sooner. But it's taking a little longer to get updated in some distributions. Therefore, we now have a blog post about this issue, complete with an explanation and workarounds: https://cockpit-project.org/blog/login-issues.html

**Reply by chall3ng3r** [reply]
Thank you, this blog post explains everything quite well.

**Comment by garrett** [comment, marked as answer]
**Update**: The official Cockpit package with the fix for the `:is():where()` login bug just landed in Ubuntu, so a normal `apt update && apt upgrade` should now work.

**Reply by chall3ng3r** [reply]
Thank you for the update.

Yes, the workaround is working and I will update as well.

Please also put together a guide to manually install v277 or whichever is the latest.

**Comment by BenoitFiset** [comment]
Hi,

I tried everything and still stuck on version 264 on focal...

checked the login.js ... is has : is(*):where(*)

```
apt update && apt upgrade
[SNIP]

sudo apt install -t focal-backports cockpit
Reading package lists... Done
Building dependency tree       
Reading state information... Done
cockpit is already the newest version (264-1ubuntu0.22.04.1~bpo20.04.1).
```
What an I missing ????

Thanks

**Reply by martinpitt** [reply]
@BenoitFiset can you please check `dpkg -l 'cockpit*'` and make sure that *all* cockpit packages are  264-1ubuntu0.22.04.1~bpo20.04.1 , in particular "cockpit-ws"?

**Reply by BenoitFiset** [reply]
Hi @martinpitt ,

here:
```
dpkg -l 'cockpit*'
Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name                   Version                         Architecture Description
+++-======================-===============================-============-===========================================
ii  cockpit                264-1ubuntu0.22.04.1~bpo20.04.1 all          Web Console for Linux servers
ii  cockpit-bridge         264-1ubuntu0.22.04.1~bpo20.04.1 amd64        Cockpit bridge server-side component
un  cockpit-dashboard      <none>                          <none>       (no description available)
ii  cockpit-doc            215-1                           all          Cockpit deployment and developer guide
ii  cockpit-machines       215-1                           all          Cockpit user interface for virtual machines
ii  cockpit-networkmanager 264-1ubuntu0.22.04.1~bpo20.04.1 all          Cockpit user interface for networking
ii  cockpit-packagekit     264-1ubuntu0.22.04.1~bpo20.04.1 all          Cockpit user interface for packages
ii  cockpit-pcp            215-1                           amd64        Cockpit PCP integration
un  cockpit-shell          <none>                          <none>       (no description available)
un  cockpit-sosreport      <none>                          <none>       (no description available)
un  cockpit-ssh            <none>                          <none>       (no description available)
ii  cockpit-storaged       264-1ubuntu0.22.04.1~bpo20.04.1 all          Cockpit user interface for storage
ii  cockpit-system         264-1ubuntu0.22.04.1~bpo20.04.1 all          Cockpit admin interface for a system
un  cockpit-systemd        <none>                          <none>       (no description available)
un  cockpit-test-assets    <none>                          <none>       (no description available)
ii  cockpit-tests          215-1                           amd64        Tests for Cockpit
un  cockpit-tuned          <none>                          <none>       (no description available)
un  cockpit-users          <none>                          <none>       (no description available)
ii  cockpit-ws             264-1ubuntu0.22.04.1~bpo20.04.1 amd64        Cockpit Web Service
```

Humm, this means I have a mismatch of 215-1 and 264-1 installed.... 

To fix, do I upgrade all the 215-1 to 264-1 and then I should be able to update all to latest ?

Thanks

**Reply by BenoitFiset** [reply]
Hi @martinpitt ,

Well went ahead and did the upgrades of 215-1 and now at this stage:

cockpit-machines wants to stay at 263-1...

```
dpkg -l 'cockpit*'
Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name                   Version                         Architecture Description
+++-======================-===============================-============-===========================================
ii  cockpit                264-1ubuntu0.22.04.1~bpo20.04.1 all          Web Console for Linux servers
ii  cockpit-bridge         264-1ubuntu0.22.04.1~bpo20.04.1 amd64        Cockpit bridge server-side component
un  cockpit-dashboard      <none>                          <none>       (no description available)
ii  cockpit-doc            264-1ubuntu0.22.04.1~bpo20.04.1 all          Cockpit deployment and developer guide
ii  cockpit-machines       263-1~bpo20.04.1                all          Cockpit user interface for virtual machines
ii  cockpit-networkmanager 264-1ubuntu0.22.04.1~bpo20.04.1 all          Cockpit user interface for networking
ii  cockpit-packagekit     264-1ubuntu0.22.04.1~bpo20.04.1 all          Cockpit user interface for packages
ii  cockpit-pcp            264-1ubuntu0.22.04.1~bpo20.04.1 amd64        Cockpit PCP integration
un  cockpit-shell          <none>                          <none>       (no description available)
un  cockpit-sosreport      <none>                          <none>       (no description available)
un  cockpit-ssh            <none>                          <none>       (no description available)
ii  cockpit-storaged       264-1ubuntu0.22.04.1~bpo20.04.1 all          Cockpit user interface for storage
ii  cockpit-system         264-1ubuntu0.22.04.1~bpo20.04.1 all          Cockpit admin interface for a system
un  cockpit-systemd        <none>                          <none>       (no description available)
un  cockpit-test-assets    <none>                          <none>       (no description available)
ii  cockpit-tests          264-1ubuntu0.22.04.1~bpo20.04.1 amd64        Tests for Cockpit
un  cockpit-tuned          <none>                          <none>       (no description available)
un  cockpit-users          <none>                          <none>       (no description available)
ii  cockpit-ws             264-1ubuntu0.22.04.1~bpo20.04.1 amd64        Cockpit Web Service
```

Try to update cockpit:

``` 
sudo apt install -t focal-backports cockpit
Reading package lists... Done
Building dependency tree
Reading state information... Done
cockpit is already the newest version (264-1ubuntu0.22.04.1~bpo20.04.1).
0 upgraded, 0 newly installed, 0 to remove and 41 not upgraded.
```

And no upgrades of cockpit :(

Next steps ?

Thanks.

**Reply by martinpitt** [reply]
That's fine, cockpit-machines is a separate source package with independent version numbers. However, your comment from above showed that you already had the latest version of cockpit-ws. Perhaps you need to force-reload the login page in your browser (Ctrl+Shift+R), it may have cached the old and broken version?

**Reply by BenoitFiset** [reply]
Hi @martinpitt,

  Humm, there is something i'm not getting here... why would a (Ctrl+Shift+R) make: ```sudo apt install -t focal-backports cockpit``` make the latest version of cockpit be able to download ..... or your telling me that _**264-1 IS the latest version**_ available for Focal !!! ???

If that's the case that 264-1 is the last version for Focal, it would be nice to state this somewhere !!

Thanks.\

**Reply by martinpitt** [reply]
It wouldn't --- your packages *are* current, -1ubuntu0.22.04.1 contains the fix. But it could be that your browser still has the login page from the older 264-1 version.

**Reply by BenoitFiset** [reply]
So 264-1ubuntu0.22.04.1~bpo20.04.1 is the latest and most current version for Focal ?

Then to have the latest version 295, I have to upgrade the OS to Jammy..... ?

**Reply by martinpitt** [reply]
Correct, yes. However, that still doesn't explain why you still see that "missing selector(:is():where()))" error on the login page -- 264-1ubuntu0.22.04.1 has the fix for that.

**Reply by BenoitFiset** [reply]
Nope that is good now.... thanks.
